packages/principled-forms/src/field/index.ts

```diff
@@ -26,15 +25,8 @@ export enum Laziness {
   Eager,
 }
 
-type ConstructorConfig<T extends InputTypes> = Required<RequiredFieldConfig<T>> | Required<OptionalFieldConfig<T>>;
-function construct<T extends InputTypes>(isRequired: boolean, config: ConstructorConfig<T>): Field<T> {
-  return isRequired ? required(config as RequiredFieldConfig<T>) : optional(config as OptionalFieldConfig<T>);
-}
-
-export function validate<T extends InputTypes, F extends Field<T>>(
-  field: F
-): F extends RequiredField<T> ? RequiredField<T> : OptionalField<T> {
-  const validities = _validate(field as Field<T>);
+export function validate<T extends InputTypes>(field: Field<T>): Field<T> {
```

Comment:
@tansongyang The users of `Field`s don't in general need to know whether it's a `Required` or an `Optional` field. You can think of these as being like your normal sum types in F♯ or Haskell: there, you wouldn't be returning `Just` or `Nothing`; you'll return `Maybe`. Same thing here: we *have* to have the independent types b/c TS' type system layers on top of JS, but we're really modeling something like a true discriminated union… just with a lot more boilerplate.

The only things that *should* need to know about the difference are the functions that map from form model to persistence model and vice versa.